### DISCUSSION TITLE
Getting an ActivityInput for an Activity in Every Workflow That has The Activity

### DISCUSSION THREAD

**Discussion by vasudan** [elsa-workflows/elsa-core#3198] [Q&A]
Hello,

Just wanted to preface this by saying I have solved this issue, I am just not sure whether it is the most correct way of doing it.

---

I have run into a problem where I need to get the value of an ActivityInput for every published workflow. 

To accomplish this I first I get every workflow blueprint that contains the activity I am looking for by iterating through every workflow provider.
```
// populated by dependency injection.
IEnumerable<IWorkflowProvider> workflowProviders

List<IWorkflowBlueprint> workflowBlueprints = new();
foreach (var workflowProvider in workflowProviders)
{
    workflowBlueprints.Add(
    workflowProvider
        .ListAsync(VersionOptions.Published)
        .Where(blueprint => blueprint.Activities
            .Any(activity => activity.Type == nameof(MyActivity))));
}
```

Afterwards I create workflow execution context for every workflow and activity execution context for every activity in that workflow and I try to get my ActivityInput.

```
// populated by dependency injection. 
ICreatesWorkflowExecutionContextForWorkflowBlueprint workflowExecutionContextFactory,
ICreatesActivityExecutionContextForActivityBlueprint activityExecutionContextFactory

List<string> activityInputs = new();
await foreach (var workflowBlueprint in workflowBlueprints)
{
    var workflowExecutionContext = await workflowExecutionContextFactory.CreateWorkflowExecutionContextAsync(workflowBlueprint, cancellationToken);
    foreach (var activityBlueprint in workflowBlueprint.Activities)
    {
        
        var activityExecutionContext = activityExecutionContextFactory.CreateActivityExecutionContext(activityBlueprint, workflowExecutionContext, cancellationToken);
        var activityAccessor = new ActivityBlueprintWrapper<MyActivity>(activityExecutionContext);
        var activityInput = await activityAccessor.EvaluatePropertyValueAsync(x => x.MyActivityInput);
        if (activityInput != null)
        {
            activityInputs.Add(activityInput);
        }
    }
}
```

My Question is, is this the best way to get ActivityInput values? This is middleware so this gets ran every single time that a request with a certain signature is made.

---

This is just about the full code of my middleware, just for a little more information.

```
public class MyMiddleware
{
    private readonly RequestDelegate _next;

    public MyMiddleware(RequestDelegate next) => _next = next;

    public async Task InvokeAsync(
        HttpContext httpContext,
        IEnumerable<IWorkflowProvider> workflowProviders,
        ICreatesWorkflowExecutionContextForWorkflowBlueprint workflowExecutionContextFactory,
        ICreatesActivityExecutionContextForActivityBlueprint activityExecutionContextFactory
        )
    {
        var path = httpContext.Request.Path.Value;

        if (path == null)
        {
            await _next(httpContext);
            return;
        }

        var cancellationToken = CancellationToken.None; 

        List<IWorkflowBlueprint> workflowBlueprints = new();
        foreach (var workflowProvider in workflowProviders)
        {
            workflowBlueprints.Add(
            workflowProvider
                .ListAsync(VersionOptions.Published)
                .Where(blueprint => blueprint.Activities
                    .Any(activity => activity.Type == nameof(MyActivity))));
        }

        List<string> activityInputs = new();
        await foreach (var workflowBlueprint in workflowBlueprints)
        {
            var workflowExecutionContext = await workflowExecutionContextFactory.CreateWorkflowExecutionContextAsync(workflowBlueprint, cancellationToken);
            foreach (var activityBlueprint in workflowBlueprint.Activities)
            {
                
                var activityExecutionContext = activityExecutionContextFactory.CreateActivityExecutionContext(activityBlueprint, workflowExecutionContext, cancellationToken);
                var activityAccessor = new ActivityBlueprintWrapper<MyActivity>(activityExecutionContext);
                var activityInput = await activityAccessor.EvaluatePropertyValueAsync(x => x.Path);
                if (activityInput != null)
                {
                    activityInputs.Add(activityInput);
                }
            }
        }
        

        // If no workflows match this path then continue.
        if (!activityPaths.Contains(path))
        {
            await _next(httpContext);
            return;
        }

        // If workflows match this path then do other stuff

    }
}
```

**Comment by vasudan** [comment]
So, after taking a look I have decided that I can probably put this logic of getting all the paths in an Elsa Event instead of in the middleware.
I created a AddPubSubConsumer that handles the TriggerIndexingFinished event that will find all the acitvity inputs using the above method.

I still am not sure whether this approach is the best idea to get all the ActivityInputs for all workflows. I also didn't see a "WorkflowAdded" or "WorkflowRemoved" event (Closest I saw was WorkflowPublished, but that didn't seem to work using the BuilderAPI) so that's why I settled with the TriggerIndexingFinished event.

**Reply by sfmskywalker** [reply]
To get all of the (evaluated) inputs, this is the way to go (although you could probably simplify it a little bit by using `IWorkflowRegistry` to get all the workflow blueprints for you and the `IWorkflowBlueprintReflector` that already deals with creating the necessary execution contexts for you).

Out of curiosity, what's the reason you need to get all of the inputs during every workflow definition update?

**Reply by vasudan** [reply]
Thanks for the reply. I'll see if I can simplify it that way.

I'm creating two WebSocket server activity. One activity 
to receive message from clients and one to broadcast a message to connected clients.

The problem is that  clients need to be able to connect to the WebSocket server before they can be broadcasted to. So, when a client tries to connect to `wss://my server.com/mypath` I need the middleware to be able to know if there is a Broadcast activity or Receive activity set up with the Path `mypath`.

Originally I had a bookmark provider setup for the Broadcast activity so I create a bookmark with the Path,  but I realize that bookmarks aren't created on startup unless the activity is the first in a workflow.

The Broadcast activity is not a triggering activity, so I can't create a bookmark for it. So to get the Path I have to use the above method.
